### PR TITLE
Add 'uniforms' Component

### PR DESCRIPTION
**Pull request by 654wak654**
**When merged this pull request will:**
- Add a small component to remove side restrictions from all vanilla uniforms.

### CODE REVIEW

**Comment by 654wak654** [comment]
Script used for generating CfgVehicles: https://gist.github.com/654wak654/130c456633d79c396f5c03ddc93ded24

**Comment by PabstMirror** [comment]
When making loadouts we often use whats looks good and don't care about side.  So there is the awkward situation where you if you take off your uniform you can't put it back on.

I think BIS's reasoning for locking uniforms was that AI are still going to do magic target IDing.

So this won't let you do "undercover" missions or let you kill a guard and take his uniform and then sneak into a base undetected (for AI).

Against humans these tactics could work, which is good but could also possibly encourage "cheap" tactics?  It isn't clear but I think using enemy uniforms may be a wacrime.  But it is something that can and has happened, so I guess I don't see any reason to allow the game to restrict it.

**Comment by jonpas** [comment]
> I think BIS's reasoning for locking uniforms was that AI are still going to do magic target IDing.

Their reasoning is one of those war conventions, they had to make the restrictions (if you remember there was no restriction in Alpha/Beta).

I think it's perfectly reasonable to do that, it's really annoying.

**Comment by commy2** [comment]
this should go into ace optionals. it would destroy TVT missions like "Coopetition" (livestream yesterday)

**Comment by bux578** [comment]
We have somthing similar for every TvT event we're doing.
No complaints

**Comment by jonpas** [comment]
> it would destroy TVT missions like "Coopetition" (livestream yesterday)

I don't see how it would destroy it?

File: addons/uniforms/$PBOPREFIX$ [hunk 1]
```diff
@@ -0,0 +1 @@
+z\ace\addons\uniforms
```
Comment (Glowbal) [inline addons/uniforms/$PBOPREFIX$]: missing new line

**Comment by commy2** [comment]
enemies taking uniforms of friendlies making them indistinguishable from friendlies.
it promotes a playstyle that is not desired by everyone.

File: addons/uniforms/$PBOPREFIX$ (hunk 1 above)
Comment (654wak654) [inline addons/uniforms/$PBOPREFIX$]: It's a PBOPREFIX ?

**Comment by 654wak654** [comment]
Of course it's possible for that to happen in TvT games, but it's also possible for it to happen IRL. You should deal with players who behave like that in some other way than making it impossible. We could just dump `vehiclelock` and make vehicles bound to factions too, stealing a vehicle can be more effective than stealing a uniform after all.

With that said, I have no problem with moving this to optionals or acex, but personally feel like this fix should be a default thing. Specially since it's already a popular piece of code in many private/public unit repositories.

**Comment by Cuel** [comment]
I'd guess there's public servers running ACE out there and I somewhat doubt they'd like this as the default behaviour

**Comment by jonpas** [comment]
Yes, let's put this into optionals @654wak654, I can see why this wouldn't be wanted as default behaviour.

**Comment by 654wak654** [comment]
Like I mentioned in slack, I don't think the optionals folder should have components like this anymore now that we have ACEX. Sooo ACEX?

**Comment by jonpas** [comment]
No, because it's a config change and still can't be toggled in-game, so it would have to be an optional in ACEX as well and we don't want optionals in both mods.

**Comment by esteldunedain** [comment]
I think moving this to ACE/Optionals is the best solution. This is one of those PBOs that some communities will love, while the general online public will probably not.

**Comment by jonpas** [comment]
You are still missing https://github.com/acemod/ACE3/pull/4435#discussion_r79310748

I am thinking a name like `remove_uniform_restrictions` might be more suitable, since it's now in optionals, previously I thought things could be added to it some day but it won't happen now, and that name would probably be better for self-explanatory reasons.

File: optionals/uniforms/functions/fnc_exportConfig.sqf [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    if !(_baseClass in (_modifyClasses apply {_x select 0})) then {
+        _baseClasses pushBackUnique _baseClass;
+    };
+    false
+} count (
+    ("!isNull (_x >> 'modelSides') &&" +
+    "{!(getArray (_x >> 'modelSides') isEqualTo [6])} &&" +
+    "{!(getArray (_x >> 'modelSides') isEqualTo [0,1,2,3])}")
+    configClasses (configFile >> "CfgVehicles")
+);
```
Comment (commy2) [inline optionals/uniforms/functions/fnc_exportConfig.sqf]: this looks confusing. what about this method?
```sqf

private _fnc_addClasses = {
    private _sides = getArray (_x >> "modelSides");

    if (!(_sides isEqualTo [6]) && {!(_sides isEqualTo [0,1,2,3])}) then {
        private _modifyClass = {
            if (!isNull (_x >> "modelSides")) exitWith {_x};
        } forEach configHierarchy _x;

        private _baseClass = inheritsFrom _modifyClass;
        _modifyClasses pushBackUnique [_modifyClass, _baseClass];

        if !(_baseClass in (_modifyClasses apply {_x select 0})) then {
            _baseClasses pushBackUnique _baseClass;
        };
    };

};

"_x call _fnc_addClasses; false" configClasses (configFile >> "CfgVehicles");
```

Comment (654wak654) [inline optionals/uniforms/functions/fnc_exportConfig.sqf]: Using `configClasses` for something other than filtering? B-but you can crash the game!!

Comment (commy2) [inline optionals/uniforms/functions/fnc_exportConfig.sqf]: Yeah. I think there is a `isNull` missing, but it looks cleaner. and does only iterate through it once for what it's worth.

**Comment by 654wak654** [comment]
Forgot to take care of this PR. What is it waiting for, the component name?

**Comment by jonpas** [comment]
A newline in pboprefix.

**Comment by 654wak654** [comment]
That's done here: https://github.com/acemod/ACE3/pull/4435/commits/a5c7ea86a7525db5d49de9d6d60b32cebc4e641b

**Comment by jonpas** [comment]
Ah, then this https://github.com/acemod/ACE3/pull/4435#issuecomment-248366402 (second paragraph).

**Comment by commy2** [comment]
You mean my comment, @jonpas? I think it can be merged even without that proposal. It was just a suggestion, but the current code is fine too.

**Comment by jonpas** [comment]
Nah, I mean rename to `remove_uniform_restrictions`.

**Comment by 654wak654** [comment]
I really feel like `remove_uniform_restrictions` is too excessive. Maybe just `uniformsides`?

**Comment by jonpas** [comment]
I think it describes exactly what the PBO does to the end user.

**Comment by 654wak654** [comment]
It does but does it have to?

**Comment by jonpas** [comment]
As we have "noactionmenu" and "nocrosshair", maybe "nouniformrestrictions" might be good. "limit" could mean something else. I am fine with either though, as long as it describes more what it does.

File: optionals/nouniformrestrictions/functions/fnc_exportConfig.sqf [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    _modifyClasses pushBackUnique [_modifyClass, _baseClass];
+    if !(_baseClass in (_modifyClasses apply {_x select 0})) then {
+        _baseClasses pushBackUnique _baseClass;
+    };
+    false
+} count (
+    ("!isNull (_x >> 'modelSides') &&" +
+    "{!(getArray (_x >> 'modelSides') isEqualTo [6])} &&" +
+    "{!(getArray (_x >> 'modelSides') isEqualTo [0,1,2,3])}")
+    configClasses (configFile >> "CfgVehicles")
```
Comment (Glowbal) [inline optionals/nouniformrestrictions/functions/fnc_exportConfig.sqf]: Can we filter out uniforms that inherit their `modelSides` value from their base class?

Comment (jonpas) [inline optionals/nouniformrestrictions/functions/fnc_exportConfig.sqf]: `configProperties` supports that IIRC.

**Pull request merged**